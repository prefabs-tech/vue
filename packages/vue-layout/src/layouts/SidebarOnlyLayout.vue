<template>
  <div class="layout sidebar-only">
    <Sidebar :menu="menu" :no-header="noHeader" class="layout-sidebar">
      <template #afterNavLinks>
        <slot name="afterNavLinks"></slot>
      </template>

      <template #afterSidebarMenu>
        <slot name="afterSidebarMenu"></slot>
        <slot name="userMenu"></slot>
        <slot name="locales">
          <LocaleSwitcher
            v-if="!noLocaleSwitcher"
            :show-badges="showBadges"
            class="locales"
          >
            <template #icon>
              <img src="../assets/svg/up-chevron.svg" alt="toggle icon" />
            </template>
          </LocaleSwitcher>
        </slot>
      </template>

      <template #footer>
        <AppFooter />
      </template>
    </Sidebar>
    <main>
      <slot></slot>
    </main>
  </div>
</template>

<script setup lang="ts">
import { useConfig } from "@prefabs.tech/vue3-config";
import { LocaleSwitcher } from "@prefabs.tech/vue3-i18n";

import AppFooter from "../components/AppFooter.vue";
import Sidebar from "../components/Sidebar.vue";

import type { SidebarMenu } from "../types";
import type { PropType } from "vue";

const { layout: layoutConfig } = useConfig();

const showBadges = layoutConfig?.localeSwitcher?.showBadges;

defineProps({
  menu: {
    required: true,
    type: Array as PropType<SidebarMenu[]>,
  },
  noLocaleSwitcher: Boolean,
  noHeader: {
    default: false,
    type: Boolean,
  },
});
</script>

<style lang="css">
@import "../assets/css/layouts/sidebarOnly.css";
</style>
