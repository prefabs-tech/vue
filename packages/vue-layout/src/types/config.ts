import type { MenuItem } from "./menu";
import type { AppConfig } from "@prefabs.tech/vue3-config";
import type { LocaleMessages, VueMessageType } from "@prefabs.tech/vue3-i18n";

interface PrefabsTechVue3LayoutConfig {
  homeRoute?: "home" | string;
  localeSwitcher: {
    showBadges: true,
  },
  logo?: string;
  logoAlt?: string;
  logoRoute?: string;
  mainMenu?: MenuItem[];
}

interface PrefabsTechVue3LayoutPluginOptions {
  config: AppConfig;
  translations?: LocaleMessages<VueMessageType>;
}

export type { PrefabsTechVue3LayoutConfig, PrefabsTechVue3LayoutPluginOptions };
