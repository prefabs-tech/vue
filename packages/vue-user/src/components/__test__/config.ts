import type { AppConfig } from "@prefabs.tech/vue3-config";
import type { LocaleMessages, VueMessageType } from "@prefabs.tech/vue3-i18n";

const config: AppConfig = {
  apiBaseUrl: "http://localhost",
  appTitle: "Test app",
  appVersion: "0.1.0",
  i18n: {
    fallbackLocale: "en",
    messages: {} as LocaleMessages<VueMessageType>,
    supportedLocales: ["en", "fr"],
  },
  features: {
    showVersion: true,
  },
  sentry: {
    dsn: "tbd",
  },
  slug: "app",
  user: {
    features: {
      confirmPassword: true,
    }
  },
  websiteDomain: "test",
};

export default config;
