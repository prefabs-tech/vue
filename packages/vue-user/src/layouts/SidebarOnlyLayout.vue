<template>
  <Layout :menu="menu" :no-locale-switcher="noLocaleSwitcher">
    <template #userMenu>
      <UserMenu v-if="showUserMenu">
        <template v-if="$slots.userMenuTrigger" #userMenuTrigger>
          <slot name="userMenuTrigger"></slot>
        </template>
      </UserMenu>
    </template>

    <template #afterNavLinks>
      <slot name="afterNavLinks"></slot>
    </template>

    <template #afterSidebarMenu>
      <slot name="afterSidebarMenu"></slot>
    </template>

    <template #locales>
      <slot name="locales"></slot>
    </template>

    <slot></slot>
  </Layout>
</template>

<script lang="ts">
export default {
  name: "SidebarOnlyLayout",
};
</script>

<script setup lang="ts">
import { useConfig } from "@prefabs.tech/vue3-config";
import { SidebarOnlyLayout as Layout } from "@prefabs.tech/vue3-layout";
import { storeToRefs } from "pinia";
import { computed } from "vue";
import { useRouter } from "vue-router";

import UserMenu from "../components/UserMenu.vue";
import useUserStore from "../store";

import type { MenuItem, SidebarMenu } from "@prefabs.tech/vue3-layout";

defineProps({
  noLocaleSwitcher: Boolean,
  showUserMenu: {
    default: true,
    required: false,
    type: Boolean,
  },
});

const userStore = useUserStore();

const { user } = storeToRefs(userStore);

const { layout: layoutConfig } = useConfig();

const router = useRouter();

const allRoutes = router.getRoutes();

const menu = computed(() => {
  let menuItems = layoutConfig?.mainMenu || [];

  menuItems = menuItems.filter((item: MenuItem) => {
    const hasValidRoute = (routeName?: string, checkAuth = true) => {
      const route = allRoutes.find((r) => r.name === routeName);

      if (!route) {
        return false;
      }

      if (!checkAuth && route.meta?.authenticated) {
        return false;
      }

      return true;
    };

    const isValidItem = (menuItem: MenuItem, checkAuth = true): boolean => {
      if (hasValidRoute(menuItem.route, checkAuth)) {
        return true;
      }

      return (
        menuItem.children?.some((child) =>
          hasValidRoute(child.route, checkAuth),
        ) ?? false
      );
    };

    return isValidItem(item, !!user.value);
  });

  return menuItems.map((item: MenuItem) => ({
    hide: item?.hide,
    icon: item?.icon,
    name: item.name,
    routeName: item.route,
    shortName: item?.shortName,
    children: item.children?.map((child: MenuItem) => ({
      hide: child?.hide,
      icon: child?.icon,
      name: child.name,
      routeName: child.route,
      shortName: child?.shortName,
    })),
  })) as SidebarMenu[];
});
</script>

<style lang="css">
.sidebar-only .footer {
  background-color: var(--sidebar-bg-color, #007aff);
  z-index: 99;
}

.sidebar-only > .sidebar .user-menu-dropdown.expanded .toggle > svg {
  transform: rotate(-180deg);
}

.sidebar-only > .sidebar .user-menu-dropdown .toggle > svg {
  transform: rotate(0);
}

.sidebar-only .user-menu > ul {
  flex-direction: column;
}

.sidebar-only .user-menu > ul > li,
.sidebar-only .user-menu-dropdown {
  --_font-size: var(--font-size-min, 0.8rem);
  --_font-weight: var(--font-weight, 450);
  --_layout-sidebar-separator-color: var(
    --layout-sidebar-separator-color,
    #dbdbdb
  );
  --dropdown-container-bg-color: #0870e5;
  --menu-highlight-color: #0870e5;

  font-size: var(--_font-size);
  font-weight: var(--_font-weight);
  width: 100%;
}

.sidebar-only > .sidebar .user-menu-dropdown {
  border-top: 1px solid var(--_layout-sidebar-separator-color);
}

.sidebar-only > .sidebar .user-menu-dropdown > .trigger {
  background-color: var(--sidebar-bg-color, #007aff);
  width: 100%;
}

.sidebar-only .user-menu > ul > li:has(.router-link-exact-active) {
  background-color: #0870e5;
}

.sidebar-only .user-menu:not(.user-menu-dropdown) > ul > li > a,
.sidebar-only .user-menu-dropdown,
.sidebar-only .user-menu-dropdown > ul.dropdown > li {
  --_height: var(--nav-menu-height, 3rem);

  align-items: center;
  display: flex;
  height: var(--_height);
  width: 100%;
}

.sidebar-only .user-menu-dropdown:hover {
  background-color: #0870e5;
}

.sidebar-only .user-menu-dropdown.expanded > ul > li:hover {
  --dropdown-bg-color-hover: #0870e5;
}

.sidebar-only .user-menu-dropdown > .dropdown {
  opacity: 0;
  transform: translate3d(0, 0, 0);
  transition:
    transform 0.3s ease,
    opacity 0.5s ease;
  width: 100%;
}

.sidebar-only .user-menu-dropdown.expanded > .dropdown {
  --dropdown-bg-color: #007aff;
  --dropdown-border: 1px solid #007aff;

  box-shadow: 0 -2px 10px 2px #0870e5;
  opacity: 1;
  transform: translate3d(0, -161.5px, 0);
  width: 100%;
}
</style>
