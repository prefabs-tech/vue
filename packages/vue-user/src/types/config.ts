import { DefineComponent } from "vue";

import type { RouteOverrides } from "./router";
import type {
  IsEmailOptions,
  StrongPasswordOptions,
} from "@prefabs.tech/vue3-form";

interface DzangolabVueUserConfig {
  apiRoutes?: {
    changeEmail?: string;
    changePassword?: string;
    getVerificationStatus?: string;
    login?: string;
    logout?: string;
    passwordReset?: string;
    passwordResetRequest?: string,
    refresh?: string;
    sendVerificationEmail?: string;
    signup?: string;
    verifyEmail?: string;
  };
  features?: {
    authProvider?: string;
    forgotPasswordResendTimeInSeconds?: number;
    loginType?: "email" | "username";
    signUp?: {
      emailVerification?: boolean;
      termsAndConditions?: {
        display?: boolean;
        route?: string;
        showCheckbox?: boolean;
      };
    };
    updateEmail?: boolean;
  };
  password?: {
    minLength: number;
  };
  options?: {
    email?: IsEmailOptions;
    password?: StrongPasswordOptions;
  };
  routes?: RouteOverrides;
  socialLogins?: string[];
  supportedRoles?: string[];
}

declare module "@prefabs.tech/vue3-config" {
  export interface AppConfig {
    user?: DzangolabVueUserConfig;
  }
}

export type { DzangolabVueUserConfig };
