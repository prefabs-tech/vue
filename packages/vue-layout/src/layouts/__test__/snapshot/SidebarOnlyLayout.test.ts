import configPlugin from "@prefabs.tech/vue3-config";
import { shallowMount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import config from "../../../components/__test__/config";
import SidebarOnlyLayout from "../../SidebarOnlyLayout.vue";

import type { VueWrapper } from "@vue/test-utils";

describe("SidebarOnlyLayout", () => {
  it("matches snapshot", () => {
    const wrapper: VueWrapper = shallowMount(SidebarOnlyLayout, {
      global: {
        plugins: [
          [
            configPlugin,
            {
              config,
            }
          ],
        ],
      },
      props: {
        menu: [
          {
            name: "Menu1",
            routeName: "menu-1",
            shortName: "M1",
          },
          {
            name: "Menu2",
            routeName: "menu-2",
            shortName: "M2",
          },
        ],
      },
      slots: {
        default: "<div class='default'></div>",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
