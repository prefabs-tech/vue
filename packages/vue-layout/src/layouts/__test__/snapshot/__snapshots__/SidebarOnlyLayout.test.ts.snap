// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarOnlyLayout > matches snapshot 1`] = `
<div
  class="layout sidebar-only"
>
  <sidebar-stub
    class="layout-sidebar"
    collapsible="true"
    menu="[object Object],[object Object]"
    noheader="false"
  />
  <main>
    
    <div
      class="default"
    />
    
  </main>
</div>
`;
